mllib/src/main/scala/org/apache/spark/ml/feature/TargetEncoder.scala

```diff
@@ -0,0 +1,447 @@
+
+  @Since("4.0.0")
+  override def load(path: String): TargetEncoder = super.load(path)
+}
+
+/**
+ * @param encodings  Original number of categories for each feature being encoded.
+ *                       The array contains one value for each input column, in order.
+ */
+@Since("4.0.0")
+class TargetEncoderModel private[ml] (
+     @Since("4.0.0") override val uid: String,
+     @Since("4.0.0") val encodings: Map[String, Map[Option[Double], Double]])
+  extends Model[TargetEncoderModel] with TargetEncoderBase with MLWritable {
+
+  @Since("4.0.0")
+  override def transformSchema(schema: StructType): StructType = {
+    inputFeatures.zip(outputFeatures)
+      .foldLeft(validateSchema(schema, fitting = false)) {
+        case (newSchema, fieldName) =>
+          val field = schema(fieldName._1)
+          newSchema.add(StructField(fieldName._2, field.dataType, field.nullable))
+      }
+  }
+
+  @Since("4.0.0")
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    validateSchema(dataset.schema, fitting = false)
+
+    // builds a column-to-column function from a map of encodings
+    val apply_encodings: Map[Option[Double], Double] => (Column => Column) =
+      (mappings: Map[Option[Double], Double]) => {
+        (col: Column) => {
+          val nullWhen = when(col.isNull,
+            mappings.get(None) match {
+              case Some(code) => lit(code)
+              case None => if ($(handleInvalid) == TargetEncoder.KEEP_INVALID) {
+                lit(mappings.get(TargetEncoder.UNSEEN_CATEGORY).get)
+              } else raise_error(lit(
+                s"Unseen null value in feature ${col.toString}. To handle unseen values, " +
+                  s"set Param handleInvalid to ${TargetEncoder.KEEP_INVALID}."))
+            })
+          val ordered_mappings = (mappings - None).toList.sortWith {
+            (a, b) => (b._1 == TargetEncoder.UNSEEN_CATEGORY) ||
+              ((a._1 != TargetEncoder.UNSEEN_CATEGORY) && (a._1.get < b._1.get))
+          }
+          ordered_mappings
+            .foldLeft(nullWhen)(
+              (new_col: Column, mapping) => {
+                val (Some(original), encoded) = mapping
+                if (original != TargetEncoder.UNSEEN_CATEGORY.get) {
+                  new_col.when(col === original, lit(encoded))
+                } else { // unseen category
+                  new_col.otherwise(
+                    if ($(handleInvalid) == TargetEncoder.KEEP_INVALID) lit(encoded)
+                    else raise_error(concat(
+                      lit("Unseen value "), col,
+                      lit(s" in feature ${col.toString}. To handle unseen values, " +
+                        s"set Param handleInvalid to ${TargetEncoder.KEEP_INVALID}."))))
+                }
+              })
+        }
+      }
+
+    dataset.withColumns(
+      inputFeatures.zip(outputFeatures).map {
+        feature =>
+          feature._2 -> (encodings.get(feature._1) match {
```

Comment:
model coefficients `encodings` stores the column names `inputCols` used in `fit`, does `encodings.get(feature._1)` requires `inputCols` in `transform` should be exactly the same as `inputCols` in `fit`?

Reply:
you're right. Fixed.
now encodings: Array[Map[Some(category), encoding]]

Reply:
btw, I´m accepting nulls in columns arrays so it's not mandatory to apply all fitted encodings. 
Find it very convenient.  Makes sense?
https://github.com/apache/spark/blob/5a67c503ce1fea57de5429ff915783d14ba0f7cf/mllib/src/main/scala/org/apache/spark/ml/feature/TargetEncoder.scala#L384 #